core/storage/filestore/impl/filesystem/filesystem_file.cpp

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "storage/filestore/impl/filesystem/filesystem_file.hpp"
+
+#include "boost/filesystem.hpp"
+#include "storage/filestore/filestore_error.hpp"
+
+using fc::storage::filestore::FileStoreError;
+using fc::storage::filestore::FileSystemFile;
+using fc::storage::filestore::Path;
+
+FileSystemFile::FileSystemFile(Path path) : path_(std::move(path)) {}
+
+Path FileSystemFile::path() const noexcept {
+  return path_;
+}
+
+fc::outcome::result<size_t> FileSystemFile::size() const noexcept {
+  OUTCOME_TRY(file_exists, exists());
+  if (!file_exists) return FileStoreError::FILE_NOT_FOUND;
+
+  return boost::filesystem::file_size(path_);
+}
+
+fc::outcome::result<void> FileSystemFile::open() noexcept {
+  OUTCOME_TRY(file_exists, exists());
+  if (!file_exists) return FileStoreError::FILE_NOT_FOUND;
+
+  fstream_.open(path_);
+
+  if (fstream_.fail()) return FileStoreError::CANNOT_OPEN;
+  return fc::outcome::success();
+}
+
+fc::outcome::result<void> FileSystemFile::close() noexcept {
+  OUTCOME_TRY(file_exists, exists());
+  if (!file_exists) return FileStoreError::FILE_NOT_FOUND;
+  if (!fstream_.is_open()) return FileStoreError::FILE_CLOSED;
+
+  fstream_.close();
+
+  if (fstream_.fail()) return FileStoreError::UNKNOWN;
+  return fc::outcome::success();
+}
+
+fc::outcome::result<size_t> FileSystemFile::read(size_t offset,
+                                                 size_t size,
+                                                 char *buffer) noexcept {
+  OUTCOME_TRY(file_exists, exists());
+  if (!file_exists) return FileStoreError::FILE_NOT_FOUND;
+  if (!fstream_.is_open()) return FileStoreError::FILE_CLOSED;
+  if (fstream_.fail()) return FileStoreError::UNKNOWN;
+
+  fstream_.seekg(offset, std::ios_base::beg);
+  fstream_.read(buffer, size);
+  auto res = fstream_.gcount();
+
+  return res;
+}
+
+fc::outcome::result<size_t> FileSystemFile::write(size_t offset,
+                                                  size_t size,
+                                                  char *buffer) noexcept {
+  OUTCOME_TRY(file_exists, exists());
+  if (!file_exists) return FileStoreError::FILE_NOT_FOUND;
+  if (!fstream_.is_open()) return FileStoreError::FILE_CLOSED;
+  if (fstream_.fail()) return FileStoreError::UNKNOWN;
+
+  fstream_.seekp(offset, std::ios_base::beg);
+  fstream_.write(buffer, size);
+  fstream_.flush();
+  auto pos = fstream_.tellp();
+  if (pos == -1) return FileStoreError::UNKNOWN;
```

Comment:
are you sure that error cannot be specified more precisely?

Reply:
Actually there are about hundred errors `in boost::system::errc` and most of them are low-level. I don't want to recount all of them.